### PR TITLE
[main][feat] transfer collateral with ext01handler

### PR DESCRIPTION
**Pull request by chenciao8**

### CODE REVIEW

File: src/handlers/Ext01Handler.sol [hunk 1]
```diff
@@ -46,13 +50,27 @@ contract Ext01Handler is OwnableUpgradeable, ReentrancyGuardUpgradeable, IExt01H
     address[] path,
     uint256 executionFee
   );
+  event LogTransferCollateral(
+    address indexed primaryAccount,
+    uint256 indexed subAccountIdFrom,
+    uint256 indexed subAccountIdTo,
```
Comment (jinlong9922) [inline src/handlers/Ext01Handler.sol:56]:
```suggestion
    uint256 indexed fromSubAccountId,
    uint256 indexed toSubAccountId,
```

File: src/handlers/Ext01Handler.sol [hunk 2]
```diff
@@ -46,13 +50,27 @@ contract Ext01Handler is OwnableUpgradeable, ReentrancyGuardUpgradeable, IExt01H
...
     address indexed primaryAccount,
     uint8 indexed subAccountId,
     uint248 amount,
     address[] path,
     uint256 minToAmount
   );
+  event LogCreateTransferCollateralOrder(
+    address indexed primaryAccount,
+    uint8 indexed subAccountIdFrom,
+    uint8 indexed subAccountIdTo,
```
Comment (jinlong9922) [inline src/handlers/Ext01Handler.sol:70]:
```suggestion
    uint8 indexed fromSubAccountId,
    uint8 indexed toSubAccountId,
```

File: src/handlers/Ext01Handler.sol [hunk 3]
```diff
@@ -181,6 +199,13 @@ contract Ext01Handler is OwnableUpgradeable, ReentrancyGuardUpgradeable, IExt01H
     uint256 minToAmount;
   }
 
+  struct CreateTransferCollateralOrderParams {
+    uint8 subAccountIdFrom;
+    uint8 subAccountIdTo;
```
Comment (jinlong9922) [inline src/handlers/Ext01Handler.sol:204]:
```suggestion
    uint8 fromSubAccountId;
    uint8 toSubAccountId;
```

File: src/handlers/Ext01Handler.sol [hunk 4]
```diff
@@ -27,6 +27,8 @@ import { ConfigStorage } from "@hmx/storages/ConfigStorage.sol";
 /// Libs
 import { HMXLib } from "@hmx/libraries/HMXLib.sol";
 
+import "forge-std/console.sol";
```
Comment (kim-hera) [inline src/handlers/Ext01Handler.sol:30]: Remove

File: src/services/CrossMarginService.sol [hunk 5]
```diff
@@ -222,22 +222,35 @@ contract CrossMarginService is OwnableUpgradeable, ReentrancyGuardUpgradeable, I
...
+
   /// @notice Calculate new trader balance after transfer collateral token.
-  /// @dev This uses to calculate new trader balance when they tranferring token as collateral.
-  /// @param _primaryAccountFrom Trader's primary address from trader's wallet to withdraw from.
-  /// @param _subAccountIdFrom Trader's Sub-Account Id to withdraw from.
-  /// @param _primaryAccountTo Trader's primary address from trader's wallet to transfer to.
-  /// @param _subAccountIdTo Trader's Sub-Account Id to transfer to.
-  /// @param _token Token that's withdrawn as collateral.
-  /// @param _amount Token withdrawing amount.
+  /// @param _params The parameters for the switch.
```
Comment (jinlong9922) [inline src/services/CrossMarginService.sol:235]:
```suggestion
  /// @param _params The parameters to transfer collateral.
```

File: src/services/interfaces/ICrossMarginService.sol [hunk 6]
```diff
@@ -47,15 +47,6 @@ interface ICrossMarginService {
     address _receiver
   ) external;
 
-  function transferCollateral(
```
Comment (jinlong9922) [inline src/services/interfaces/ICrossMarginService.sol:50]: why remove

File: src/services/CrossMarginService.sol [hunk 7]
```diff
@@ -222,22 +222,35 @@ contract CrossMarginService is OwnableUpgradeable, ReentrancyGuardUpgradeable, I
...
   function transferCollateral(
-    address _primaryAccountFrom,
-    uint8 _subAccountIdFrom,
-    address _primaryAccountTo,
-    uint8 _subAccountIdTo,
-    address _token,
-    uint256 _amount
-  ) external nonReentrant onlyWhitelistedExecutor onlyAcceptedToken(_token) {
+    TransferCollateralParams calldata _params
```
Comment (jinlong9922) [inline src/services/CrossMarginService.sol:237]: why we changed the way to pass parameters

**Review by jinlong9922 — CHANGES_REQUESTED** [review]

**Review by john-xina-p88 — APPROVED** [review]

File: src/services/CrossMarginService.sol (hunk 7 above)
Comment (chenciao8) [reply src/services/CrossMarginService.sol:237]: prevent stack too deep

File: src/services/interfaces/ICrossMarginService.sol (hunk 6 above)
Comment (chenciao8) [reply src/services/interfaces/ICrossMarginService.sol:50]: re-added

File: src/services/interfaces/ICrossMarginService.sol [hunk 8]
```diff
@@ -23,6 +23,16 @@ interface ICrossMarginService {
     uint256 minToAmount;
   }
 
+
+  struct TransferCollateralParams {
+    address fromPrimaryAccount;
+    uint8 fromSubAccountId;
+    address toPrimaryAccount;
```
Comment (jinlong9922) [inline src/services/interfaces/ICrossMarginService.sol:30]: do we allow to transfer between account ?

**Review by jinlong9922 — CHANGES_REQUESTED** [review]

File: src/services/interfaces/ICrossMarginService.sol (hunk 8 above)
Comment (chenciao8) [reply src/services/interfaces/ICrossMarginService.sol:30]: yes, for future support

**Review by jinlong9922 — APPROVED** [review]

**Pull request merged**